pkg/cloud-init/cloud-init.go

```diff
@@ -456,124 +468,149 @@ func GenerateLocalData(vmiName string, namespace string, data *CloudInitData) er
+	err = ioutil.WriteFile(dataSource.UserFile, userData, 0644)
 	if err != nil {
 		return err
 	}
-	err = ioutil.WriteFile(metaFile, metaData, 0644)
+	err = ioutil.WriteFile(dataSource.MetaFile, dataSource.MetaData, 0644)
 	if err != nil {
 		return err
 	}
 
 	files := make([]string, 0, 3)
-	files = append(files, metaFile)
-	files = append(files, userFile)
+	files = append(files, dataSource.MetaFile)
+	files = append(files, dataSource.UserFile)
 
 	if len(networkData) > 0 {
-		err = ioutil.WriteFile(networkFile, networkData, 0644)
+		err = ioutil.WriteFile(dataSource.NetworkFile, networkData, 0644)
 		if err != nil {
 			return err
 		}
-		files = append(files, networkFile)
+		files = append(files, dataSource.NetworkFile)
 	}
 
 	switch data.DataSource {
 	case DataSourceNoCloud:
-		err = cloudInitIsoFunc(isoStaging, "cidata", dataBasePath)
+		err = cloudInitIsoFunc(dataSource.IsoStaging, "cidata", dataSource.DataBasePath)
 	case DataSourceConfigDrive:
-		err = cloudInitIsoFunc(isoStaging, "config-2", dataBasePath)
+		err = cloudInitIsoFunc(dataSource.IsoStaging, "config-2", dataSource.DataBasePath)
 	}
 	if err != nil {
 		return err
 	}
-	diskutils.RemoveFile(metaFile)
-	diskutils.RemoveFile(userFile)
-	diskutils.RemoveFile(networkFile)
+	diskutils.RemoveFile(dataSource.MetaFile)
+	diskutils.RemoveFile(dataSource.UserFile)
+	diskutils.RemoveFile(dataSource.NetworkFile)
 
-	if err := diskutils.DefaultOwnershipManager.SetFileOwnership(isoStaging); err != nil {
+	if err := diskutils.DefaultOwnershipManager.SetFileOwnership(dataSource.IsoStaging); err != nil {
 		return err
 	}
 
-	err = os.Rename(isoStaging, iso)
+	err = os.Rename(dataSource.IsoStaging, dataSource.Iso)
 	if err != nil {
-		log.Log.Reason(err).Errorf("Cloud-init failed to rename file %s to %s", isoStaging, iso)
+		log.Log.Reason(err).Errorf("Cloud-init failed to rename file %s to %s", dataSource.IsoStaging, dataSource.Iso)
 		return err
 	}
 
-	log.Log.V(2).Infof("generated nocloud iso file %s", iso)
+	log.Log.V(2).Infof("generated nocloud iso file %s", dataSource.Iso)
 	return nil
 }
 
+func generateLocalDataSource(vmiName string, namespace string, data *CloudInitData) (*LocalDataSource, error) {
+	var err error
+	ds := &LocalDataSource{}
+	ds.DomainBasePath = getDomainBasePath(vmiName, namespace)
+	ds.DataBasePath = filepath.Join(ds.DomainBasePath, "data")
+
+	if data.DataSource == DataSourceNoCloud {
+		ds, err = generateLocalDataSourceNoCloudData(vmiName, namespace, data, ds)
+		if err != nil {
+			return nil, err
+		}
+	} else if data.DataSource == DataSourceConfigDrive {
+		ds, err = generateLocalDataSourceConfigDrive(vmiName, namespace, data, ds)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("Invalid cloud-init data source: '%v'", data.DataSource)
+	}
+	return ds, nil
+}
+
+func generateLocalDataSourceNoCloudData(vmiName string, namespace string, data *CloudInitData, ds *LocalDataSource) (*LocalDataSource, error) {
```

Comment:
Why do you need to return `ds`  if you passed a pointer to it?  
After second though about naming - it seems that hese two functions just Set or Populate the DataSource , and the caller function is the one creating/generating the dataSorce